NMS-13760: Split SNMP Property Extenders into multiple pages (#4086)
OpenNMS/opennms: docs/modules/operation/pages/performance-data-collection/snmp-property-extenders/snmp-interface.adoc

> == SNMP Interface Property Extender

Reviewer: ```suggestion
= SNMP Interface Property Extender
```

The SNMP Interface property extender does much the same job as the Pointer-Like Index property extender, but it is specialized for importing properties from the `ifTable`.
Resources representing rows in the `ifTable` are modeled differently in {page-component-title} compared to other tabular resource types, and this extender accounts for those differences.

> === When to Use the SNMP Interface Property Extender

Reviewer: ```suggestion
== When to Use the SNMP Interface Property Extender
```

Use the SNMP Interface property extender when the string property you want to import is associated with a network interface which is represented by a row in the `ifTable`.
For example, the `dot1dBasePortTable` has its own index which does not share any components with any other table, but its `dot1dBasePortIfIndex` column contains a value that is a valid `ifIndex`.

[source, snmp-mib]
----
   dot1dBasePortEntry OBJECT-TYPE
       -- ...
       DESCRIPTION
           "A list of information for each port of the bridge."
       -- ...
       INDEX  { dot1dBasePort } <1>
  -- ...

   dot1dBasePortIfIndex OBJECT-TYPE <2>
       -- ...
       DESCRIPTION
           "The value of the instance of the ifIndex object,
           defined in IF-MIB, for the interface corresponding
           to this port."
       ::= { dot1dBasePortEntry 2 }
  -- ...

ifEntry OBJECT-TYPE
    -- ...
    DESCRIPTION
            "An entry containing management information applicable to a
            particular interface."
    INDEX   { ifIndex } <3>
    ::= { ifTable 1 }
----

> <1> The entry type for `dot1dBasePortTable` is indexed on `dot1dBasePort`, which has no significance outside this table

Reviewer 2: ```suggestion
<1> The entry type for `dot1dBasePortTable` is indexed on `dot1dBasePort`, which has no significance outside this table.
```

> <2> But `dot1dBasePortTable` contains column `dot1dBasePortIfIndex`, which tells us the `ifIndex` corresponding to the physical port underlying to the associated bridge base port

Reviewer 2: ```suggestion
<2> But `dot1dBasePortTable` contains column `dot1dBasePortIfIndex`, which tells us the `ifIndex` corresponding to the physical port underlying the associated bridge base port.
```

> <3> `ifIndex` is the index of the `ifTable` entry type (and also of the `ifXTable` entry type)

Reviewer 2: ```suggestion
<3> `ifIndex` is the index of the `ifTable` entry type (and also of the `ifXTable` entry type).
```

By using this extender, it's possible to import string attributes from the `ifTable`, `ifXTable`, or another table that augments the `ifTable`.

> === Configuring the SNMP Interface Property Extender

Reviewer: ```suggestion
== Configuring the SNMP Interface Property Extender
```

The SNMP Interface property extender expects two or three parameters:

.SNMP Interface Property Extender Parameters
[options="header", cols="1,3,2"]
|===
| Name
| Description
| Default value

3+| *Required*

| source-attribute

> | The alias name of the string property to be imported from the source MIB table.

Reviewer 2: ```suggestion
| The alias name of the string property to import from the source MIB table.
```

    <property instance="dot1dBasePortEntry" alias="dot1dBasePortIfDescr" class-name="org.opennms.netmgt.collectd.InterfaceSnmpPropertyExtender"> <3>
      <parameter key="source-ifindex-attribute" value="interfaceIndex"/>
      <parameter key="source-attribute" value="interfaceDescr"/> <4>
      <parameter key="target-ifindex-pointer-column" value="dot1dBasePortIfIndex"/>
    </property>
    <property instance="dot1dBasePortEntry" alias="dot1dBasePortIfName" class-name="org.opennms.netmgt.collectd.InterfaceSnmpPropertyExtender"> <5>
      <parameter key="source-ifindex-attribute" value="interfaceIndex"/>
      <parameter key="source-attribute" value="interfaceName"/> <6>
      <parameter key="target-ifindex-pointer-column" value="dot1dBasePortIfIndex"/>
    </property>
    <property instance="dot1dBasePortEntry" alias="dot1dBasePortIfAlias" class-name="org.opennms.netmgt.collectd.InterfaceSnmpPropertyExtender"> <7>
      <parameter key="source-ifindex-attribute" value="interfaceIndex"/>
      <parameter key="source-attribute" value="interfaceAlias"/> <8>
      <parameter key="target-ifindex-pointer-column" value="dot1dBasePortIfIndex"/>
    </property>
  </group>
</groups>
----

> <1> First we collect all of `ifIndex`, `ifDescr`, `ifName`, and `ifAlias` in a group associated with the `ifIndex` source resource-type, using modified names to avoid collisions with internal workings (the `ifIndex` type is built in, so we do not need a custom resource-type definition for it)

Reviewer 2: ```suggestion
<1> First we collect all of `ifIndex`, `ifDescr`, `ifName`, and `ifAlias` in a group associated with the `ifIndex` source resource-type, using modified names to avoid collisions with internal workings (the `ifIndex` type is built in, so we do not need a custom resource-type definition for it).
```

> <2> Then we collect the pointer-like column `dot1dBasePortIfIndex` in the target group

Reviewer 2: ```suggestion
<2> Then we collect the pointer-like column `dot1dBasePortIfIndex` in the target group.
```

> <3> To derive the `dot1dBasePortIfDescr` string property, we tell the extender which target attribute contains the pointer-like value, which source column needs to have a matching value, and that we want to "pull over" the `interfaceDescr` property

Reviewer 2: ```suggestion
<3> To derive the `dot1dBasePortIfDescr` string property, we tell the extender which target attribute contains the pointer-like value, which source column needs to have a matching value, and that we want to use the `interfaceDescr` property.
```

Reviewer 2: Once again, if "pull over" is more accurate, then ignore "use".

Author: I don't think it makes much difference. In German it sounds also better with `use`. :+1:

> <4> from the source group

Reviewer 2: ```suggestion
<4> From the source group
```

> <5> Deriving `dot1dBasePortIfName` is almost identical, except that we want the property `interfaceName`

Reviewer 2: ```suggestion
<5> Deriving `dot1dBasePortIfName` is almost identical, except that we want the property `interfaceName`.
```

> <6> from the source group instead

Reviewer 2: ```suggestion
<6> From the source group instead
```

> <7> Again with `dot1dBasePortIfAlias`, we repeat ourselves except that our desired property from the source group is `interfaceAlias` <8>

Reviewer 2: <8> should be on the next line.
